posts/2020/03/pytorch-tensoriterator-internals.md

```diff
@@ -0,0 +1,273 @@
+`TensorIterator` performs optimizations to make sure that at least
+most of the iterations happen on contiguos data to take advantage of hierarchical cache-based
+memory architectures (think dimension coalescing and reordering for maximum data locality).
+
+Now a multi-dimensional tensor will have multiple stride values depending on the dimension
+you want to iterate over, so `TensorIterator` will directly compute the strides that
+get passed into the loop by
+by itself within the `build()` function. How exactly it computes the dimension
+to iterate over is something that should be properly understood in order to use `TensorIterator`
+effectively.
+
+If you're performing a reduction operation (see the sum code in [ReduceOps.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ReduceOps.cpp#L384)),
+`TensorIterator` will figure out the dimensions that will be reduced depending
+on the shape of the input and output tensor, which determines how the input will be broadcast
+over the output. If you're
+performing a simple pointwise operation between two tensors (like a `addcmul` from 
+[PointwiseOps.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/PointwiseOps.cpp#L31))
+the iteration will happen over the entire tensor, without providing a choice of the dimension.
+This will allow TensorIterator to freely parallelize the computation, without guarantees of
+the order of execution (since it does not matter anyway).
+
+For something like a cumulative sum operation, where you want be able to choose the dimension
+to reduce but iterate over multiple non-reduced dimensions (possibly in parallel), you
+must first re-stride the tensors, and then use these tensors 
+for creating a `TensorIterator`. In order to understand how this bit works, lets go over
+the code for the [kernel](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp#L21) that executes the [cumsum](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp#L71) function.
+
+The important bits of this function are like so:
+
+``` cpp
+auto self_sizes = ensure_nonempty_vec(self.sizes().vec());
+self_sizes[dim] = 1;
+
+auto result_restrided = restride_dim(result, dim, self_sizes);
+auto self_restrided = restride_dim(self, dim, self_sizes);
+
+auto iter = TensorIterator();
+iter.dont_compute_common_dtype();
+iter.dont_resize_outputs();
+iter.add_output(result_restrided);
+iter.add_input(self_restrided);
+iter.build();
+```
+You can see that we first change the size of the tensors to `1` on the
+reduction dimension so that the dimension collapsing logic inside
+`TensorIterator#build` will know which dimension to broadcast the result
+over. Setting the dimension in this way is akin to telling `TensorIterator`
+to skip the dimension. We then restride the tensors using `restride_dim` and
+then use the restrided tensors for building the `TensorIterator`.
```

Comment:
How is broadcasting is connected to skipping the dimension in here? Since you set size 1 for all inputs/outputs, you do not broadcast, but, well, say TensorIterator to skip that dimension.

Reply:
You can set any size for inputs/outputs, then TensorIterator with check whether it can come up with a common broadcasted size...